microsoft/testsuites/cvm/cvm_boot.py

```diff
@@ -145,3 +157,23 @@ def verify_boot_success_after_component_upgrade(
                 pcrs_after_reboot,
                 "PCR values changed even though no boot component was updated",
             ).is_equal_to(pcrs_before_reboot)
+
+    def _add_remote_repo(self, node: RemoteNode, repo_url: str, log: Logger) -> None:
+        posix_os: Posix = cast(Posix, node.os)
+        posix_os.add_repository(repo_url)
+
+    def _create_local_repo(
+        self, node: RemoteNode, source_tarball: Path, log: Logger
+    ) -> None:
+        working_path = Path(node.get_working_path())
+        tarball_file = source_tarball.name
+        tarball_path = working_path / tarball_file
+
+        # copy tarball to remote
+        result = node.tools[RemoteCopy].copy_to_remote(
+            src=source_tarball,
+            dest=working_path,
+        )
+        log.debug(f"tarball copied to: {result}")
+
+        node.tools[CreateRepo].create_repo_from_tarball(tarball_path)
```

Comment:
How about call install_packages of the os?

Reply:
I think at least for tdnf/dnf, to install a local rpm, you'd have to give the whole filename of the rpm, and that allows you to install an older version of a package as well.

Reply:
Done